### PR TITLE
Improve AI pathfinding

### PR DESCRIPTION
**Pull request by Morlic-fo**
This PR replaces the current AI pathfinding with a new pathfinding algorithm that considers (re)fuel mechanics and allows to find paths that require detours to resupply (and planned for the future around systems blocked by e.g. a stationary monster).

The core algorithm is based on A* with the ```universe.shortestPathDistance```as heuristic. However, this implementation keeps track of the fleet's fuel during pathfinding. Reached nodes are only closed for a certain fuel level at that node, i.e. longer paths that leave the fleet with more fuel at a system will be considered in pathfinding. This is how detours for refueling are found.

The previous algorithms used jump distance rather than starlane distance, consider that fact when comparing results. Nowadays, the AI won't stop at each system anymore so using ```universe.shortestPathDistance``` is the prefered metric as long as fuel is taken into account.

With regard to performance:
1) The ```universe.shortestPathDistance``` as A* heuristic is exact if no detour is required. This means the shortest path will be found without traversing other branches in that case.
2) The algorithm ignores paths that would be "much" longer than the ```universe.shortestPathDistance``` to lessen the impact of worst-case scenarios (i.e. no valid path available). "Much longer" in this PR is defined as either twice as long as the ```universe.shortestPathDistance``` or at least 2000uu longer.

If it still turns out to be a performance hazard in some specific situations, there are some more optimizations that could be implemented (mentioned in ToDos) but performance seemed fine in my tests so far, so I didn't bother yet. As last measure, reimplementation in C++ is another option.

### CODE REVIEW

File: default/python/AI/pathfinding.py
Comment (Dilvish-fo) [inline default/python/AI/pathfinding.py]: If there is no reason to ever work with empty paths here (and I don't see any), then the final item in ```path_info.path``` will always be the same as ```current``` and I would like to suggest we keep the heap structure simpler and then here say
```python
(_, path_info) = heappop(queue)
current = path_info.path[-1]
```

Although it adds an extra line right here, in several other places (wherever we construct new entries for the queue) it keeps things simpler.   And to me it seems a bit helpful that the start of the loop has this explicit reminder of the relationship between ```path_info.path``` and ```current```.

**Review by Dilvish-fo — COMMENTED** [review]
This is looking very good.  I just thought of a couple minor-ish things that hadn't come to mind in our previous review.

File: default/python/AI/pathfinding.py
Comment (Morlic-fo) [reply default/python/AI/pathfinding.py]: I was talking about the refueling mechanics if the fleet is stationary in an unsupplied system. I clarified that comment.

If our path reaches the target system but is e.g. 0.1 fuel off, then it could be prefered to let it wait there for one turn rather than having to take a detour. The logic to find suitable circumstances isn't trivial and needs some thought but it would be useful in (various) edge cases especially for those living organic ships.

File: default/python/AI/pathfinding.py
Comment (Morlic-fo) [reply default/python/AI/pathfinding.py]: Good call. I changed it.

**Comment by Dilvish-fo** [comment]
I've been holding off from merging this to give CJ a chance to review this final version, but he had already given comments on most of it before, and so unless you have some plans to add more for review, @Morlic-fo , I'd suggest it's waited long enough and we go ahead and merge.

File: default/python/AI/pathfinding.py
Comment (Cjkjvfnby) [inline default/python/AI/pathfinding.py:77]: > Thus, the target system must be at most maxFuel + 1 jumps away in order to reach the system under standard conditions.

`standard conditions` need definition.  `maxFuel + 1 jumps `  is distance that can be reached on worst case  (we start on the edge of the our borders and go to outer space.)

File: default/python/AI/pathfinding.py
Comment (Cjkjvfnby) [inline default/python/AI/pathfinding.py]: `u` ???

File: default/python/AI/pathfinding.py
Comment (Cjkjvfnby) [inline default/python/AI/pathfinding.py]: No need parenthesis here.

File: default/python/AI/pathfinding.py
Comment (Cjkjvfnby) [inline default/python/AI/pathfinding.py:121]: `current in supplied_systems` means `fleet.maxFuel - 1`

File: default/python/AI/pathfinding.py
Comment (Cjkjvfnby) [inline default/python/AI/pathfinding.py]: Please add comment how enemy fleets and monsters is counted during pathfinding.

File: default/python/AI/pathfinding.py
Comment (Cjkjvfnby) [inline default/python/AI/pathfinding.py]: 2000 is magic constant, pleas give it meaningful name, and may be some comments.

**Review by Cjkjvfnby — CHANGES_REQUESTED** [review]
I have checked comments and docstrings.   

I am not common with A* need time to understand it enough to make code review for algorithm implementation.   I am a little busy with my study, so please don't wait me.

File: default/python/AI/pathfinding.py
Comment (Dilvish-fo) [reply default/python/AI/pathfinding.py]: Currently this doesn't do anything about enemy fleets and monsters (nor does the function this is replacing).  I've been experimenting, though, with some augmentation for this that does take them into account, I plan to post a PR for that shortly after this gets merged.

File: default/python/AI/pathfinding.py
Comment (Morlic-fo) [reply default/python/AI/pathfinding.py]: It is a relatively stupid query but implies no strict logic error. For example, function ```foo``` calls this function to determine whether a scout can reach the target system and then return. A query that triggers this check and the answer to it are both perfectly fine and valid in this context and definitely not an ```error```.

Should it already be caught by the caller? Probably so. That's why it is logged as a warning.

File: default/python/AI/pathfinding.py
Comment (Dilvish-fo) [reply default/python/AI/pathfinding.py]: The comment you quote is about the process of building the set of intermediates explored while determining the final result, which is clearly stated to be the shorted path meeting the fuel requirements.  The two statements are not in conflict.   Though I suppose some more words could be added to help clarify the distinction.

File: default/python/AI/pathfinding.py
Comment (Morlic-fo) [reply default/python/AI/pathfinding.py]: It avoids a potential issue and the additional code is trivial.

File: default/python/AI/pathfinding.py
Comment (Dilvish-fo) [reply default/python/AI/pathfinding.py]: Right, Morlic, if this wasn't simply a typo slip from "resupply-sideroutes" then please consider it as a clarification.

File: default/python/AI/pathfinding.py
Comment (Morlic-fo) [reply default/python/AI/pathfinding.py]: Open for discussion but for me:
```a, b = foo()``` implies ```foo()``` returns 2 values by default (i.e. an iterable of exactly 2 items)
```(a, b) = foo()``` implies ```foo()``` returns 1 value by default but that value is a tuple in this particular use case.

File: default/python/AI/pathfinding.py
Comment (Dilvish-fo) [reply default/python/AI/pathfinding.py]: Not necessarily stupid, though it could be valid to require the fuel check be done prior to calling this function.

Suppose the AI is checking several fleets to see which one, if any, can make it to the target with the desired remaining fuel, do we necessarily want to force the AI to separately check the maxfuel levels before calling this method, or can we simply let this method handle that aspect?    

Seems to me to be fine, slightly simpler, to let it be handled here.

File: default/python/AI/pathfinding.py
Comment (Morlic-fo) [reply default/python/AI/pathfinding.py]: Added a TODO how I have envisioned to handle this - an additional argument for systems that are not to be considered for pathfinding. The logic what those systems are (e.g. those blocked by monsters) is supposed to be figured out by the calling function in advance.

File: default/python/AI/pathfinding.py
Comment (Morlic-fo) [reply default/python/AI/pathfinding.py:121]: One would intuitively think so. I initially missed this as well.

However, as Dilvish correctly pointed out and as you can easily verify ingame by moving a fleet out of supply, the first jump moving out of supply does not consume any fuel (or is immediately refueled).

File: default/python/AI/pathfinding.py
Comment (Dilvish-fo) [reply default/python/AI/pathfinding.py]: My recollection of the code is that fuel-refilling occurs during the movement stage, prior to any meter changes.  If a system becomes newly-supplied on a given turn (whether from meter change or from a blockade being lifted that same turn), then yes, a fleet in such a supplied system be at fuel less than max (and this is consistent with my recollection from experience).  So no, I don't think that simplification works.

File: default/python/AI/pathfinding.py
Comment (Morlic-fo) [reply default/python/AI/pathfinding.py]: I rephrased the comment. Is this clearer now or does it need more work?

File: default/python/AI/pathfinding.py
Comment (Dilvish-fo) [reply default/python/AI/pathfinding.py]: p.s., I recently added [an Order of Turn Events section to the Scripting Tutorial](http://www.freeorion.org/index.php/FOCS_Scripting_Tutorial#Order_of_Events_During_the_Turn) to help clarify such things for everyone.

File: default/python/AI/pathfinding.py
Comment (Dilvish-fo) [reply default/python/AI/pathfinding.py:121]: No, there is no fuel cost to depart from a supplied system.

edit: whoops, Morlic's comment didn't show up before I submitted my comment, to clarify, my "no' was to cj, not to Morlic.

Comment (Cjkjvfnby) [reply default/python/AI/pathfinding.py:121]: I suggest to add comment about this into code.

File: default/python/AI/pathfinding.py
Comment (Cjkjvfnby) [reply default/python/AI/pathfinding.py]: > returns 2 values by default

There is no 2 values. It is always one value  `return a, b` is `return (a, b)`.  Tuple is `iterable of exactly 2 items`  in this particular use case.

**Review by Cjkjvfnby — APPROVED** [review]
There is one comment left  about docstring.

**Review by Dilvish-fo — APPROVED** [review]

**Pull request merged**
